controlplane/kubeadm/controllers/upgrade.go

```diff
@@ -105,9 +106,40 @@ func (r *KubeadmControlPlaneReconciler) upgradeControlPlane(
 		return ctrl.Result{}, err
 	}
 
+	if rolloutStrategyIsSet(kcp) && controlPlaneIsSetToScaleDown(kcp) {
+		if status.Nodes >= *kcp.Spec.Replicas {
+			return r.scaleDownControlPlane(ctx, cluster, kcp, controlPlane, machinesRequireUpgrade)
+		}
+
+		if status.Nodes < *kcp.Spec.Replicas {
+			return r.scaleUpControlPlane(ctx, cluster, kcp, controlPlane)
+		}
+	}
+
 	if status.Nodes <= *kcp.Spec.Replicas {
 		// scaleUp ensures that we don't continue scaling up while waiting for Machines to have NodeRefs
 		return r.scaleUpControlPlane(ctx, cluster, kcp, controlPlane)
 	}
 	return r.scaleDownControlPlane(ctx, cluster, kcp, controlPlane, machinesRequireUpgrade)
 }
```

Comment:
I'm not sure if we have to worry about checking if it's set, since there is defaulting in place (I believe that the conversion webhook will also trigger the defaulting webhook when going from v1alpha3 to v1alpha4, if I'm not mistaken).

I think things can be simplified a bit further if this used the values defined in the rollout strategy:
```suggestion
	// RolloutStrategy is currently defaulted and validated to be RollingUpdate
	// We can ignore MaxUnavailable because we are enforcing health checks before we get here.
	maxNodes = kcp.Spec.RolloutStrategy.RollingUpdate.MaxSurge.IntValue() + *kcp.Spec.Replicas
	
	if status.Nodes < maxNodes {
		// scaleUp ensures that we don't continue scaling up while waiting for Machines to have NodeRefs
		return r.scaleUpControlPlane(ctx, cluster, kcp, controlPlane)
	}
	return r.scaleDownControlPlane(ctx, cluster, kcp, controlPlane, machinesRequireUpgrade)
}
```